File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/usecase/UpdatePlanUseCase.kt
package com.nogeass.passalarm.domain.usecase

import com.nogeass.passalarm.domain.model.AlarmPlan
import com.nogeass.passalarm.domain.repository.AlarmPlanRepository
import javax.inject.Inject

class UpdatePlanUseCase @Inject constructor(
    private val planRepository: AlarmPlanRepository,
    private val reschedule: RescheduleNextNUseCase
) {
    suspend fun execute(plan: AlarmPlan) {
        planRepository.save(plan.copy(updatedAt = System.currentTimeMillis()))
        reschedule()
    }
}

File: apps/ios/PassAlarm/Domain/UseCase/UpdatePlanUseCase.swift
import Foundation

final class UpdatePlanUseCase: Sendable {
    private let planRepository: AlarmPlanRepositoryProtocol
    private let reschedule: RescheduleNextNUseCase

    init(planRepository: AlarmPlanRepositoryProtocol,
         reschedule: RescheduleNextNUseCase) {
        self.planRepository = planRepository
        self.reschedule = reschedule
    }

    func execute(_ plan: AlarmPlan) async throws {
        var updated = plan
        updated.updatedAt = Date()
        try await planRepository.save(updated)
        try await reschedule.execute()
    }
}
